ops/main.py

```diff
@@ -356,9 +325,255 @@ def _should_use_controller_storage(db_path: Path, meta: CharmMeta) -> bool:
+
+        charm = charm_type(framework)
+        dispatcher.ensure_event_links(charm)
+        return charm
+
+    @staticmethod
+    def _setup_model_backend():
+        # this is a standalone method to facilitate overriding by subclassing in testing
+        return ops.model._ModelBackend()
+
+    @staticmethod
+    def _setup_root_logging(backend: ops.model._ModelBackend):
+        debug = "JUJU_DEBUG" in os.environ
+        setup_root_logging(backend, debug=debug)
+
+        # our hello world
+        logger.debug(
+            "ops %s up and running.",
+            ops.__version__,
+        )  # type:ignore
+
+    def _setup_storage(self, dispatcher: _Dispatcher):
+        charm_state_path = self.charm_spec.charm_root / self.CHARM_STATE_FILE
+
+        use_juju_for_storage = self._use_juju_for_storage
+        if use_juju_for_storage and not ops.storage.juju_backend_available():
+            # raise an exception; the charm is broken and needs fixing.
+            msg = 'charm set use_juju_for_storage=True, but Juju version {} does not support it'
+            raise RuntimeError(msg.format(JujuVersion.from_environ()))
+
+        if use_juju_for_storage is None:
+            use_juju_for_storage = _should_use_controller_storage(
+                charm_state_path,
+                self._charm_meta
+            )
+        elif use_juju_for_storage:
+            warnings.warn("Controller storage is deprecated; it's intended for "
+                          "podspec charms and will be removed in a future release.",
+                          category=DeprecationWarning)
+
+        if use_juju_for_storage:
+            if dispatcher.is_restricted_context():
+                # TODO: jam 2020-06-30 This unconditionally avoids running a collect metrics event
+                #  Though we eventually expect that Juju will run collect-metrics in a
+                #  non-restricted context. Once we can determine that we are running
+                #  collect-metrics in a non-restricted context, we should fire the event as normal.
+                logger.debug('"%s" is not supported when using Juju for storage\n'
+                             'see: https://github.com/canonical/operator/issues/348',
+                             dispatcher.event_name)
+                # Note that we don't exit nonzero, because that would cause Juju to rerun the hook
+                exit(0)
+
+        if self._use_juju_for_storage:
+            store = ops.storage.JujuStorage()
+        else:
+            store = ops.storage.SQLiteStorage(charm_state_path)
+        return store
+
+    def _setup_framework(
+            self,
+            dispatcher: _Dispatcher,
+            model_backend: "ops.model._ModelBackend"
+    ):
+        meta = self._charm_meta
+        model = ops.model.Model(meta, model_backend)
+        store = self._setup_storage(dispatcher)
+        framework = ops.framework.Framework(store, self.charm_spec.charm_root, meta, model)
```

Comment:
I'm not familiar enough with the main/framework code to know this and haven't taken the time to walk through things to figure it out: does not passing in the event name change anything here? It seems like maybe it would impact detecting custom events? We have the dispatcher in this method, so it seems like it would be simple to pass it like the existing code does, unless there's a specific reason to not do that.

Reply:
oh, I missed that. Good catch.